source/server/server.h

```diff
@@ -252,6 +259,8 @@ class InstanceImpl : Logger::Loggable<Logger::Id::main>, public Instance {
   Envoy::MutexTracer* mutex_tracer_;
   Http::ContextImpl http_context_;
   std::unique_ptr<Memory::HeapShrinker> heap_shrinker_;
+  std::thread::id main_thread_id_;
+  std::unordered_map<Stage, std::vector<StageCallback>> stage_callbacks_;
```

Comment:
`absl::flat_hash_map`? Or perhaps more pragmatically, just use a vector since the possible stages are all known at compile time, and there are only two.

Reply:
From a quick search it seems like std::unordered_map is much more common than absl::flat_hash_map in the envoy code base. The Stage enum values are strongly typed so to use them as indexes into a vector I think I'd need to do a static_cast to int which seems somewhat ugly so using a map seems better.

Reply:
Most of the code in Envoy by lines was written before `absl::flat_hash_map` was open sourced, so I suspect that's the main reason why it's more common. I think it's still a good practice to use `flat_hash_map` where possible in new code, so I'd use it here.

I think it's good as a map rather than a vector, your reasoning makes total sense. This isn't performance critical so if it's more readable this way that's the right solution.